shaders/raycast_shader.frag

```diff
@@ -0,0 +1,221 @@
+#version 330 
+out vec4 fragColor; // Final output color
+smooth in vec3 vUV; // 3D texture coordinates form vertex shader interpolated by rasterizer
+vec4 vFragColor=vec4(0);
+
+
+uniform sampler2D lut;// transferfunction
+uniform sampler2D volume_2D;// volume dataset
+uniform sampler3D volume_3D;// volume dataset
+uniform sampler2D depth;
+uniform sampler3D blendVolume;// volume dataset
+uniform vec3 camPos;
+uniform vec3 clip_min;
+uniform vec3 clip_max;
+uniform bool useBlend;
+uniform bool texture_atlas;
+uniform bool useLut;
+uniform bool clipping;
+uniform bool useMultiLut;
+uniform int channel;
+uniform float slices;
+uniform float dimension;
+uniform float blendAlpha;
+uniform float threshold;
+uniform float multiplier;
+uniform mat4 P_inv;
+uniform mat4 clipPlane;
+
+
+vec2 intersect_box(vec3 orig, vec3 dir, vec3 clip_min, vec3 clip_max) { 
+	//"dir += vec3(0.0001); 
+	vec3 inv_dir = 1.0 / dir;
+	vec3 tmin_tmp = (clip_min - orig) * inv_dir;
+	vec3 tmax_tmp = (clip_max - orig) * inv_dir;
+	vec3 tmin = min(tmin_tmp, tmax_tmp);
+	vec3 tmax = max(tmin_tmp, tmax_tmp);
+	float t0 = max(tmin.x, max(tmin.y, tmin.z));
+	float t1 = min(tmax.x, min(tmax.y, tmax.z)); 
+	return vec2(t0, t1); 
+}
+
+vec4 sample_model(sampler2D volume_texture, vec2 start_position, vec2 end_position, float ratio)
+{
+    // Sample model texture as 3D object, alpha is initialized as the max channel
+    vec4 model_sample = mix (texture(volume_texture, start_position),texture(volume_texture, end_position),ratio);
+	model_sample.a = max(model_sample.r, max(model_sample.g, model_sample.b));
+    if(model_sample.a < 0.20) model_sample.a *= 0.1;
+    // Sample transfer texture
+    return texture(lut, vec2(clamp(model_sample.a, 0.0, 1.0))); 
+	
+}
+     
+
+void main()
+{
+		vec3 dataPos = vUV;
+		// get the object space position by subracting 0.5 from the
+        // 3D texture coordinates. Then subtraact it from camera position
+        // and normalize to get the ray marching direction
+        vec3 geomDir = normalize( dataPos - camPos);
+		
+		// get the t values for the intersection with the box
+		vec2 t_hit = intersect_box(camPos, geomDir,clip_min,clip_max);
+		
+		// We don't want to sample voxels behind the eye if it's
+		// inside the volume, so keep the starting point at or in front
```

Comment:
Tab size goes back and forth a lot? I know if you're using VS Code you can do `ALT+SHIFT+f` to do a quick format

Reply:
oh yes. I missed formatting the shaders on vs code. visual studio didnt want to format those files.